### PR TITLE
BUG: UsageInputs.validate - handle optional

### PR DESCRIPTION
**Pull request by thermokarst**
~TODO:~

- [x] add unit tests
- [x] make sure that `DiagnosticUsage` works
- [x] make sure that `ArtifactAPIUsage` works

### CODE REVIEW

File: qiime2/sdk/usage.py [hunk 1]
```diff
@@ -64,8 +64,20 @@ def __repr__(self):
     def validate(self, signature):
         provided = set(self.values.keys())
         inputs, params = signature.inputs, signature.parameters
-        exp_inputs = {k for k, v in inputs.items() if not v.has_default()}
-        exp_params = {k for k, v in params.items() if not v.has_default()}
+
```
Comment (thermokarst) [inline qiime2/sdk/usage.py]: The heart of the PR is in this hunk.

**Comment by thermokarst** [comment]
ccing @David-Rod - I think @ebolyen mentioned to me that you might've come across this issue in your own Usage API work.

File: qiime2/sdk/usage.py [hunk 2]
```diff
@@ -85,11 +98,12 @@ def build_opts(self, signature, scope):
 
         for name, signature in signature.signature_order.items():
             if name in self.values:
-                if isinstance(self.values[name], ScopeRecord) \
-                        and self.values[name].ref in scope.records:
-                    value = self.values[name].result
+                v = self.values[name]
```
Comment (thermokarst) [inline qiime2/sdk/usage.py:101]: This change isn't strictly related to this PR, but it was bugging me.

File: qiime2/sdk/usage.py [hunk 3]
```diff
@@ -85,11 +98,12 @@ def build_opts(self, signature, scope):
...
+                if isinstance(v, ScopeRecord) \
+                        and v.ref in scope.records:
```
Comment (andrewsanchez) [inline qiime2/sdk/usage.py:103]: This can be on one line now.

**Comment by thermokarst** [comment]
I am pretty confused about the flake8 errors on travis...

**Comment by thermokarst** [comment]
Okay, squared away travis.

**Pull request merged**